py/dml/codegen.py

```diff
@@ -394,6 +395,462 @@ def memoization_common_prelude(name, site, outp, throws, make_ref):
                      make_ref('ran', TInt(8, True)),
                      mkCompound(site, unrun + has_run + running))]
 
+class TypeSequenceInfo:
+    """Bookkeeping surrounding a realtyped unique type sequence"""
+    def __init__(self, types, uniq):
+        self.types = types
+        self.uniq = uniq
+        self.after_on_hooks = {}
+        self.struct = (TStruct({f'comp{i}': typ
+                                 for (i, typ) in enumerate(types)},
+                                label=f'_typeseq_{self.uniq}')
+                       if types else None)
+        type_keys = ', '.join(f"'{typ.key()}'" for typ in types)
+        self.string_key = f'({type_keys})'
+
+    def get_after_on_hook(self, aoh_prim_key, param_to_msg_comp, no_params,
+                          createnew=False):
+        param_to_msg_comp_key = tuple(param_to_msg_comp.get(i)
+                                      for i in range(no_params))
+        try:
+            return self.after_on_hooks[(aoh_prim_key, param_to_msg_comp_key)]
+        except KeyError:
+            if not createnew:
+                raise
+            info = after_on_hook_info_constructors[type(aoh_prim_key)](
+                self, aoh_prim_key, param_to_msg_comp)
+            self.after_on_hooks[(aoh_prim_key, param_to_msg_comp_key)] = info
+            return info
+
+class TypeSequenceInfoDict:
+    """Mappings from a unique type sequence to its TypeSequenceInfo"""
+    def __init__(self):
+        self._dict = {}
+        self._list = []
+
+    def get(self, index, realtyped=False, createnew=False):
```

Comment:
`create_new`

Reply:
Also I don't like this signature, the classname announces "dict" but it differs a lot from dict.get. Perhaps rename to `lookup` to mark the difference? Or create separate `__contains__` method for the  `create_new=False` case? Or shoehorn into what a `collections.defaultdict` use would look like?

Reply:
I _refuse_ to complicate this by toying around with Pythons magic methods or the classes of `collections`. They are so horribly badly documented to the point it's impossible to know whether you're doing things _right_. I'd _much much much_ rather just rename this class if you really do hate the name that much.

Reply:
Now irrelevant.